### PR TITLE
Enhance LPM integration with architecture support and utilities

### PR DESCRIPTION
**Pull request by filipelautert**
## Impact

- [ ] Bug fix (non-breaking change which fixes expected existing functionality)
- [x] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 

## Description

This PR introduces the following enhancements:

- **LPM Integration:**
  - Enhanced `LpmCommandStep` by adding test coverage and support for the `--download` argument.
  - Improved installation logic across platforms using the GitHub API.
  - Expanded functionality with added error handling and fallback mechanisms.

- **Utility Methods:**
  - Added `fetchAsString` method in `DownloadUtil` to facilitate URL content retrieval with better error handling and custom configurations.

- **Expanded OS Architecture Detection:**
  - New support for the `s390x` architecture on Linux systems.

- **General Improvements:**
  - Updated logging and messaging for clarity.
  - Resolved Windows version download issues and removed unnecessary error messages.

These updates enhance user experience, platform compatibility, and overall system utility in handling LPM-related tasks.

## Things to be aware of

- Refinements to the GitHub API integration have been properly implemented to ensure smooth functionality.
- Comprehensive unit tests have been added for new and modified features.

## Things to worry about

- Ensure rigorous testing across all platforms to identify potential edge cases.
- Review the seamlessness of LPM command-line functionality after new additions.

## Additional Context

These changes aim to improve the command-line utility experience for users and enhance system-wide compatibility while supporting advanced use cases.

### CODE REVIEW

**Comment by StevenMassaro** [comment]
Is there a ticket associated with this?

**Comment by filipelautert** [comment]
> Is there a ticket associated with this?

@StevenMassaro yes, https://datical.atlassian.net/browse/DAT-20511 . I reused my old poc branch and forgot to rename it.

File: liquibase-cli/src/main/java/liquibase/integration/commandline/LiquibaseCommandLine.java [hunk 1]
```diff
@@ -455,6 +455,79 @@ public int execute(String[] args) {
...
+        
+        // Known flags that should not be passed to LPM binary - discovered via reflection
+        Set<String> knownFlags = discoverLpmCommandFlags();
+
+        for (String arg : strings) {
+            if (lpmArgs == null) {
+                returnArgs.add(arg);
+                
+                if ("lpm".equals(arg.toLowerCase())) {
+                    foundInit = true;
```
Comment (MalloD12) [inline liquibase-cli/src/main/java/liquibase/integration/commandline/LiquibaseCommandLine.java]: I guess this one should be deleted now.

Comment (filipelautert) [reply liquibase-cli/src/main/java/liquibase/integration/commandline/LiquibaseCommandLine.java]: Oh right, removing the unused variable

**Review by MalloD12 — APPROVED** [review]
LGTM now.

File: liquibase-cli/src/main/java/liquibase/integration/commandline/LiquibaseCommandLine.java [hunk 2]
```diff
@@ -455,6 +455,77 @@ public int execute(String[] args) {
...
+            Scope.getCurrentScope().setLpmArgs(lpmArgs.toString().trim());
+        }
+        return returnArgs.toArray(new String[0]);
+    }
+
+    /**
+     * Discovers CommandArgumentDefinition fields from LpmCommandStep using reflection
+     * and returns their names with "--" prefix.
+     */
+    private Set<String> discoverLpmCommandFlags() {
```
Comment (StevenMassaro) [inline liquibase-cli/src/main/java/liquibase/integration/commandline/LiquibaseCommandLine.java:496]: Aren't there existing factory methods that you could use here?

File: liquibase-standard/src/main/java/liquibase/command/core/LpmCommandStep.java [hunk 3]
```diff
@@ -0,0 +1,561 @@
...
+        try {
+            Scope.getCurrentScope().getLog(getClass()).fine("Executing LPM command: " + String.join(" ", lpmExecArgs));
+            final Process process = new ProcessBuilder(lpmExecArgs).redirectErrorStream(true).start();
+            InputStream stdOut = process.getInputStream();
+            try (BufferedReader stdOutReader = new BufferedReader(new InputStreamReader(stdOut))) {
+                String line;
+                while ((line = stdOutReader.readLine()) != null) {
+                    Scope.getCurrentScope().getUI().sendMessage(line);
+                }
+            }
```
Comment (StevenMassaro) [inline liquibase-standard/src/main/java/liquibase/command/core/LpmCommandStep.java]: There are Apache Commons methods  that you might be able to use here instead

File: liquibase-standard/src/main/java/liquibase/command/core/LpmCommandStep.java [hunk 4]
```diff
@@ -0,0 +1,561 @@
...
+        for (int i = 0; i < maxLength; i++) {
+            int v1 = i < parts1.length ? Integer.parseInt(parts1[i]) : 0;
+            int v2 = i < parts2.length ? Integer.parseInt(parts2[i]) : 0;
+            
+            if (v1 > v2) return true;
+            if (v1 < v2) return false;
+        }
+        
+        return false; // versions are equal
+    }
```
Comment (StevenMassaro) [inline liquibase-standard/src/main/java/liquibase/command/core/LpmCommandStep.java]: Can you use the Java 9 version class here to do this comparison? https://docs.oracle.com/javase/9/docs/api/java/lang/module/ModuleDescriptor.Version.html

File: liquibase-standard/src/main/java/liquibase/command/core/LpmCommandStep.java [hunk 5]
```diff
@@ -0,0 +1,561 @@
...
+        
+        // Determine platform for download URL
+        String platform = determinePlatform();
+        final String lpmUrl = String.format("https://github.com/liquibase/liquibase-package-manager/releases/download/v%s/lpm-%s-%s.zip", version, version, platform);
+        final String lpmZip = String.format("%s%slpm.zip", lpmHome, File.separator);
+        
+        Scope.getCurrentScope().getUI().sendMessage("Downloading LPM version " + version + " for " + platform + " from " + lpmUrl);
+        
+        try {
+            Path downloadedFilePath = DownloadUtil.downloadToFile(lpmUrl, new File(lpmZip));
```
Comment (StevenMassaro) [inline liquibase-standard/src/main/java/liquibase/command/core/LpmCommandStep.java]: Do you want to download this to a temp file?

File: liquibase-standard/src/main/java/liquibase/command/core/LpmCommandStep.java [hunk 6]
```diff
@@ -0,0 +1,561 @@
...
+                    throw new UnexpectedLiquibaseException("Could not create directory: " + parentDir.getAbsolutePath());
+                }
+                
+                // Extract file
+                try (FileOutputStream fos = new FileOutputStream(newFile)) {
+                    int len;
+                    while ((len = zis.read(buffer)) > 0) {
+                        fos.write(buffer, 0, len);
+                    }
+                }
```
Comment (StevenMassaro) [inline liquibase-standard/src/main/java/liquibase/command/core/LpmCommandStep.java]: this should be using a library method

File: liquibase-standard/src/main/java/liquibase/util/DownloadUtil.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+public class DownloadUtil {
```
Comment (StevenMassaro) [inline liquibase-standard/src/main/java/liquibase/util/DownloadUtil.java]: What methods in this class are actually being used?

File: liquibase-standard/src/main/java/liquibase/util/DownloadUtil.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+                StreamUtil.copy(response, file);
+            }
+            log.fine("Saved " + url + " to " + pathAsFile.getAbsolutePath());
+
+
+        } catch (Exception e) {
+            throw new UnexpectedLiquibaseException("Error downloading from " + url + ": " + e.getMessage(), e);
+        }
+        return pathAsFile.toPath();
+    }
```
Comment (StevenMassaro) [inline liquibase-standard/src/main/java/liquibase/util/DownloadUtil.java]: Can't you do this with commons?

File: liquibase-standard/src/main/java/liquibase/util/DownloadUtil.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+                    }
+                    return response.toString();
+                }
+            } else {
+                throw new UnexpectedLiquibaseException("HTTP request failed with response code: " + responseCode);
+            }
+        } catch (Exception e) {
+            throw new UnexpectedLiquibaseException("Error fetching content from " + url + ": " + e.getMessage(), e);
+        }
+    }
```
Comment (StevenMassaro) [inline liquibase-standard/src/main/java/liquibase/util/DownloadUtil.java]: Can't you do this with commons?

File: liquibase-standard/src/main/java/liquibase/util/DownloadUtil.java (hunk 7 above)
Comment (filipelautert) [reply liquibase-standard/src/main/java/liquibase/util/DownloadUtil.java]: This class was moved from `src/test` , so all of the methods are in use.

File: liquibase-standard/src/main/java/liquibase/util/DownloadUtil.java (hunk 8 above)
Comment (filipelautert) [reply liquibase-standard/src/main/java/liquibase/util/DownloadUtil.java]: yes, I can!

File: liquibase-standard/src/main/java/liquibase/util/DownloadUtil.java (hunk 7 above)
Comment (StevenMassaro) [reply liquibase-standard/src/main/java/liquibase/util/DownloadUtil.java]: I don't love having code that is only used by tests shipped with production code. Maybe moving this class won't be necessary once you start using commons methods.

**Review by StevenMassaro — APPROVED** [review]

File: liquibase-standard/src/main/java/liquibase/util/DownloadUtil.java (hunk 7 above)
Comment (filipelautert) [reply liquibase-standard/src/main/java/liquibase/util/DownloadUtil.java]: Yeap, commons solved it all 🥳
